scripts/maintenance/standard-header-guards.pl

```diff
@@ -0,0 +1,57 @@
+#!/usr/bin/perl
+
+## Copyright (C) 1996-2023 The Squid Software Foundation and contributors
+##
+## Squid software is distributed under GPLv2+ license and includes
+## contributions from numerous individuals and organizations.
+## Please see the COPYING and CONTRIBUTORS files for details.
+##
+
+# source-maintenance plugin to standardize the header guards
+
+use strict;
+use warnings;
+
+my $filename = $ARGV[0];
+
+if ($filename !~ /\.h$/i || $filename =~ /^scripts\/boilerplate.h/i) {
+    while (<>) { print; }
+    exit(0);
+}
+
+my @accumulate=();
+my $current_guard = undef;
+
+my $first_ifdef_pos = undef;
+my $last_ifdef_pos = undef;
+while (<>) {
+    if (!defined($first_ifdef_pos) && /^#ifndef\s+(\w+_H)/ ) {
+        $current_guard = $1;
+        $first_ifdef_pos = $#accumulate+1;
+    } elsif (/^#endif/) {
+        $last_ifdef_pos = $#accumulate+1;
+    }
+    push(@accumulate, $_);
+}
+
+die("cannot detect header guard in $filename") unless defined($current_guard);
+die("no #endif in $filename") unless defined($last_ifdef_pos);
+die ("#define $current_guard doesn't immediately follow first #ifndef in $filename")
+    unless ($accumulate[$first_ifdef_pos+1] =~ /^#define\s+$current_guard/);
```

Comment:
The current PR code is more complex/tricky than it needs to be, but I do not insist on simplifying it further. In this particular case, the complexity may lead to overread of the accumulate array though:

```suggestion
    unless ($first_ifdef_pos < $#accumulate && $accumulate[$first_ifdef_pos+1] =~ /^#define\s+$current_guard/);
```

Reply:
Overread can only happen if the first ifndef is the last line in the file (and if it is, source formatting is the last of our problems and this is likely ).

There are therefore only two cases where this can happen:
- no endif/truncated file - this is caught by the test in line 38
- last endif is before the first ifdef - I'm adding an explicit test for it
- I will add an explicit test for this scenario, with a more useful error message; I would expect it will never ever trigger tho